### PR TITLE
Handle error gracefully when we can't retrieve an image

### LINKED ISSUE #250: Comparing metadata for images that don't exist locally causes panic due to SIGSEGV
## Observed

When running `container-diff` using one or more images that do not exist locally and passing `--type=metadata`, there is a panic :O

```bash
container-diff diff cats dogs --type=metadata 
2018/07/25 15:31:31 No matching credentials found for index.docker.io, falling back on anonymous
2018/07/25 15:31:31 No matching credentials found for index.docker.io, falling back on anonymous
WARN[0001] Diff may be inaccurate: open : no such file or directory 
WARN[0001] Diff may be inaccurate: open : no such file or directory 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x81f612]

goroutine 1 [running]:
github.com/GoogleContainerTools/container-diff/differs.getMetadataList(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/GoogleContainerTools/container-diff/differs/metadata_diff.go:77 +0x42
github.com/GoogleContainerTools/container-diff/differs.getMetadataDiff(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/GoogleContainerTools/container-diff/differs/metadata_diff.go:62 +0x8a
github.com/GoogleContainerTools/container-diff/differs.MetadataAnalyzer.Diff(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/GoogleContainerTools/container-diff/differs/metadata_diff.go:40 +0xb2
github.com/GoogleContainerTools/container-diff/differs.(*MetadataAnalyzer).Diff(0xc7ff18, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	<autogenerated>:1 +0xab
github.com/GoogleContainerTools/container-diff/differs.DiffRequest.GetDiff(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/GoogleContainerTools/container-diff/differs/differs.go:62 +0x1b8
github.com/GoogleContainerTools/container-diff/cmd.diffImages(0x7ffd24b21fc0, 0x4, 0x7ffd24b21fc5, 0x4, 0xc42028fce0, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/GoogleContainerTools/container-diff/cmd/diff.go:110 +0x601
github.com/GoogleContainerTools/container-diff/cmd.glob..func4(0xc57a60, 0xc4202f4330, 0x2, 0x3)
	/go/src/github.com/GoogleContainerTools/container-diff/cmd/diff.go:46 +0x89
github.com/GoogleContainerTools/container-diff/vendor/github.com/spf13/cobra.(*Command).execute(0xc57a60, 0xc4202f42a0, 0x3, 0x3, 0xc57a60, 0xc4202f42a0)
	/go/src/github.com/GoogleContainerTools/container-diff/vendor/github.com/spf13/cobra/command.go:766 +0x2c1
github.com/GoogleContainerTools/container-diff/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc57cc0, 0x0, 0x0, 0x0)
	/go/src/github.com/GoogleContainerTools/container-diff/vendor/github.com/spf13/cobra/command.go:852 +0x334
github.com/GoogleContainerTools/container-diff/vendor/github.com/spf13/cobra.(*Command).Execute(0xc57cc0, 0x1f, 0x0)
	/go/src/github.com/GoogleContainerTools/container-diff/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
	/go/src/github.com/GoogleContainerTools/container-diff/main.go:35 +0x66
```

## Expected
There should be an error message probably, but not a panic/runtime exception, e.g. without `--type=metadata` the output is:

```bash
➜  kaniko git:(local_integration_tests) container-diff diff cats dogs                
2018/07/25 15:33:08 No matching credentials found for index.docker.io, falling back on anonymous
2018/07/25 15:33:08 No matching credentials found for index.docker.io, falling back on anonymous
WARN[0001] Diff may be inaccurate: open : no such file or directory 
WARN[0001] Diff may be inaccurate: open : no such file or directory 
ERRO[0001] Error getting diff with AptAnalyzer: stat : no such file or directory 
ERRO[0001] Could not retrieve diff: Could not perform diff on {<nil>   []} and {<nil>   []} 
```
Interestingly, the output also doesn't match - but at least doesn't panic! - when using `--type=file`:

```bash
➜  kaniko git:(local_integration_tests) container-diff diff cats dogs --type=file    
2018/07/25 15:33:35 No matching credentials found for index.docker.io, falling back on anonymous
2018/07/25 15:33:35 No matching credentials found for index.docker.io, falling back on anonymous
WARN[0001] Diff may be inaccurate: open : no such file or directory 
WARN[0001] Diff may be inaccurate: open : no such file or directory 

-----File-----

These entries have been added to : None

These entries have been deleted from : None

These entries have been changed between  and : None
```

## Steps to reproduce
Assuming you do not have the images `cat` and `dog` locally (or run `docker system prune -a` so you can use ANY image name XD), run:

```bash
container-diff diff cats dogs --type=metadata
```

_This is with version `v0.11.0`._

### PR DESCRIPTION
**Pull request by nkubala**
This fixes the error handling for processing images before we do the diff or analysis. Errors are propagated back through a channel to the main loop, which will only execute the diff/analysis if we didn't receive any errors back from the image processing. Error handling also happens in the main loop now, so we'll get consistent results across diff types.

This also changes a bunch of error messages to add some more context when we do receive an error.

```
➜ ./out/container-diff diff dogs gcr.io/google-appengine/python:latest
2018/07/26 12:27:12 No matching credentials found for index.docker.io, falling back on anonymous
ERRO[0001] error retrieving image dogs: UNAUTHORIZED: "authentication required" 
➜ ./out/container-diff analyze dogs                                   
2018/07/26 12:27:16 No matching credentials found for index.docker.io, falling back on anonymous
ERRO[0000] error retrieving image dogs: UNAUTHORIZED: "authentication required"
```

Fixes #250

### CODE REVIEW

File: cmd/diff.go [hunk 1]
```diff
@@ -69,33 +70,60 @@ func checkFilenameFlag(_ []string) error {
 	return errors.New("Please include --types=file with the --filename flag")
 }
 
+func processImage(imageName string, imageMap map[string]*pkgutil.Image, wg *sync.WaitGroup, errChan chan<- error) {
```
Comment (bobcatfish) [inline cmd/diff.go]: Is there a name for this function that might be a bit more specific to what this function is doing? I wasn't sure what `processImage` meant when I was reading this. 

It seems like the purpose of this function is to look up sometthing about the image (im not sure what tho - are we actually getting the bytes of the image itself, or looking up some metadata or something else?) and put that something into `imageMap`

So some other ideas:
- `AddImageToMap`
- `GetMetadataForImage`
- `FindImageSource`
- `DownloadImage`

(You might have some other name ideas tho since you probably know what exactly `getImageForName` is doing)

File: cmd/diff.go [hunk 2]
```diff
@@ -69,33 +70,60 @@ func checkFilenameFlag(_ []string) error {
...
-			image, err := getImageForName(imageName)
-			imageMap[imageName] = &image
-			if err != nil {
-				logrus.Warningf("Diff may be inaccurate: %s", err)
-			}
-		}(imageArg, imageMap)
+	img2, ok := imageMap[image2Arg]
+	if !ok {
+		return fmt.Errorf("cannot find image %s", image2Arg)
 	}
```
Comment (bobcatfish) [inline cmd/diff.go]: what would it mean if processImage returned for both images without errors, but one or both of these checks failed?

File: cmd/diff.go [hunk 3]
```diff
@@ -69,33 +70,60 @@ func checkFilenameFlag(_ []string) error {
...
+		err, ok := <-errChan
+		if !ok {
+			break
+		}
+		errs = append(errs, err.Error())
+	}
+
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, "\n"))
+	}
```
Comment (bobcatfish) [inline cmd/diff.go]: this logic to collect together errors into one error seems like a good candidate for being moved into a separate function

File: cmd/diff.go [hunk 4]
```diff
@@ -69,33 +69,58 @@ func checkFilenameFlag(_ []string) error {
 	return errors.New("Please include --types=file with the --filename flag")
 }
 
+func processImage(imageName string, imageMap map[string]*pkgutil.Image, wg *sync.WaitGroup, errChan chan<- error) {
+	defer wg.Done()
+	image, err := getImageForName(imageName)
+	if image.Image == nil {
```
[the remark this answers is not in the document]
Comment (nkubala) [reply cmd/diff.go]: Ok this logic isn't great so I'm gonna rework it a bit. Digging into the go-containerregistry code, if we get an error returned back from any of the three sources (tar/daemon/remote), the image returned will always be `nil`. So it's equivalent to just change this back to a vanilla `if err != nil` check.

The reason I was checking for `image.Image == nil` is to avoid putting the image into the returned map so we didn't attempt to clean up an empty path, but that logic is already happening in `pkgutil.CleanupImage` so this is safe. I moved the deferred calls up before we return the errors, and we're good.

TLDR answers for your questions: 1) yes, and 2) not possible

File: cmd/diff.go (hunk 1 above)
Comment (nkubala) [reply cmd/diff.go]: This function is essentially just a concurrent wrapper around `getImageForName`, which retrieves a reference to an image and a reader for the actual bytes of the image. I can change this to `getImage` (the `imageMap` isn't really a useful struct here so I don't want to refer to it), but the reason I called it `processImage` is because `getImageForName` actually does a bit of image processing, i.e. unpacks the image filesystem to the local filesystem. I should probably just add some comments explaining what each function does :)

File: cmd/diff.go (hunk 2 above)
Comment (nkubala) [reply cmd/diff.go]: removed these since this logic is duplicated elsewhere

File: cmd/diff.go (hunk 3 above)
Comment (nkubala) [reply cmd/diff.go]: agreed, done

File: cmd/diff.go [hunk 5]
```diff
@@ -69,33 +69,58 @@ func checkFilenameFlag(_ []string) error {
...
+		errChan <- fmt.Errorf("error retrieving image %s: %s", imageName, err.Error())
```
[the remark this answers is not in the document]
Comment (nkubala) [reply cmd/diff.go]: technically the format types don't match (`string` vs `error`), but this seems to work so I'll change it back

File: cmd/analyze_test.go [hunk 6]
```diff
@@ -29,11 +29,11 @@ var analyzeArgNumTests = []testpair{
 func TestAnalyzeArgNum(t *testing.T) {
 	for _, test := range analyzeArgNumTests {
 		err := checkAnalyzeArgNum(test.input)
-		if (err == nil) != test.expected_output {
-			if test.expected_output {
-				t.Errorf("Got unexpected error: %s", err)
-			} else {
+		if (err == nil) != test.shouldError {
```
Comment (bobcatfish) [inline cmd/analyze_test.go]: nice, `shouldError` is a way better name!

File: cmd/diff.go [hunk 7]
```diff
@@ -66,50 +67,74 @@ func checkFilenameFlag(_ []string) error {
...
+
 	wg.Wait()
+	close(errChan)
 
 	if noCache && !save {
 		defer pkgutil.CleanupImage(*imageMap[image1Arg])
 		defer pkgutil.CleanupImage(*imageMap[image2Arg])
 	}
 
+	if err := readErrorsFromChannel(errChan); err != nil {
```
Comment (bobcatfish) [inline cmd/diff.go]: niiiiiiiice

File: cmd/root.go [hunk 8]
```diff
@@ -131,6 +132,9 @@ func checkIfValidAnalyzer(_ []string) error {
 	return nil
 }
 
+// getImageForName infers the source of an image and retrieves a v1.Image reference to it.
+// Once a reference is obtained, it attempts to unpack the v1.Image's reader's contents
+// into a temp directory on the local filesystem.
```
Comment (bobcatfish) [inline cmd/root.go]: ty ty, makes sense now!
